### PR TITLE
WIP: portforward: symmetric half-open connection logic for client and kubelet

### LINKED ISSUE #27680: [k8s.io] Port forwarding [k8s.io] With a server that expects a client request should support a client that connects, sends data, and disconnects [Conformance] {Kubernetes e2e suite}
https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/kubernetes-e2e-gke-staging-parallel/4404/

Failed: [k8s.io] Port forwarding [k8s.io] With a server that expects a client request should support a client that connects, sends data, and disconnects [Conformance] {Kubernetes e2e suite}

```
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/portforward.go:263
Jun 18 20:37:24.831: Expected "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx" from server, got "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
```

### PR DESCRIPTION
**Pull request by sttts**
This PR adds support for half-open connections in both directions to the port-forwarding code in the client (used by `kubectl port-foward`) and the container runtimes in the kubelet.

The layers between those two ends of a port-forward use htpstreams (using SPDY right now, but soon also websockets: https://github.com/kubernetes/kubernetes/pull/33684) instead of raw TCP. This PR is purely about tunneling TCP in SPDY (or soon websockets).

TODO:
- [x] probably fix some unit tests
- [ ] add e2e tests for the server closing a connection early

Fixes https://github.com/sttts/kubernetes/commit/ed021fed4c38b5c82696fea84b3173e98bc16922#commitcomment-19876346.

This is in the context of https://github.com/kubernetes/kubernetes/issues/27680 which might be a race with closed connections somewhere. Though it seems that this very change will not fix #27680 as the former is about early client close, the later about early server close.

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/37072)

### CODE REVIEW

**Comment by ncdc** [comment]
I think this LGTM. cc @liggitt 

@k8s-bot gci gke e2e test this

**Comment by sttts** [comment]
@k8s-bot gci gke e2e test this

File: pkg/client/unversioned/portforward/portforward.go [hunk 1]
```diff
@@ -203,8 +203,13 @@ func (pf *PortForwarder) listenOnPortAndAddress(port *ForwardedPort, protocol st
 
 // getListener creates a listener on the interface targeted by the given hostname on the given port with
 // the given protocol. protocol is in net.Listen style which basically admits values like tcp, tcp4, tcp6
-func (pf *PortForwarder) getListener(protocol string, hostname string, port *ForwardedPort) (net.Listener, error) {
-	listener, err := net.Listen(protocol, fmt.Sprintf("%s:%d", hostname, port.Local))
+func (pf *PortForwarder) getListener(protocol string, hostname string, port *ForwardedPort) (*net.TCPListener, error) {
+	addr, err := net.ResolveTCPAddr(protocol, fmt.Sprintf("%s:%d", hostname, port.Local))
```
Comment (sttts) [inline pkg/client/unversioned/portforward/portforward.go]: fixed

File: pkg/util/httpstream/httpstream.go [hunk 2]
```diff
@@ -147,3 +152,90 @@ func Handshake(req *http.Request, w http.ResponseWriter, serverProtocols []strin
...
+	go func() {
+		defer close(copiesDone)
+		copiesGroup.Wait()
+	}()
+
+	// wait for either both copies finish or an error occurs
+	select {
+	case <-copiesDone:
+	case err := <-localError: // catch the first error, potentially ignore the second
+		return err
```
Comment (sttts) [inline pkg/util/httpstream/httpstream.go]: If both go routines error out, `copiesDone` could be closed. So we could miss the error, I guess.

Comment (fraenkel) [inline pkg/util/httpstream/httpstream.go]: We can fix that by not using defer.

**Pull request closed without merging**